rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,411 @@
+a time-based loop s.t. between 1 and 2 upgrade attempts are made each hour.
+
+When the teleport agent is healthy, it should periodically publish a "schedule" indicating when it is appropriate to
+attempt an upgrade (schedule variants are discussed later). Updaters should obey the emitted schedule to ensure that
+any necessary downtime occurs during a maintenance window.
+
+Abstractly, the updater loop looks like this:
+
+```
+┌─► Loop ◄──────────────────────────┐
+│     │                             │
+│     │                             │
+│     ▼                             │
+│   Critical Update?                │
+│     │   │                         │
+│     │   │                         │
+│    no   └──────yes────────┐       │
+│     │                     │       │
+│     │                     │       │
+│     ▼                     ▼       │
+│   Agent Healthy?     Attempt Upgrade
+│     │  │                      ▲   ▲
+│     │  │                      │   │
+│    yes └────────no────────────┘   │
+│     │                             │
+no    │                            yes
+│     ▼                             │
+└───Within Maintenance Window? ─────┘
+```
+
+Note that this is a bit vague on timing/frequency. Different updaters use different scheduling/looping mechanics (discussed
+in more detail below).
+
+Different variants will require different upgrade window specifications. In order to minimize client-side logic, the auth server
+will present agents with the raw "schedule spec" for the agent's given deployment context. For k8s this will be a json blob
+understood by the controller and exported to a kubernetes config object. For VM/bare metal this will be a systemd
+`OnCalendar` string (e.g. `Tue,Wed *-*-* 00:00:00 UTC`) to be written out to a systemd `override.conf`.
+
+```proto3
+message GetAgentMaintenanceScheduleRequest {
+    string TeleportVersion = 1;
+    AgentUpgraderKind Kind = 2;
+}
+
+enum AgentUpgraderKind {
+    KUBE_CONTROLLER = 1;
+    SYSTEMD_TIMER = 2;
+}
+
+message GetAgentMaintenanceScheduleResponse {
+    string Spec = 1;
+}
+
+service AuthServer {
+    // ...
+
+    rpc GetAgentMaintenanceSchedule(GetAgentMaintenanceScheduleRequest) returns (GetAgentMaintenanceScheduleResponse);
+}
+
+// note: opting for pull-based over push-based since agents will want to discover the window immediately on
+// startup under some conditions, but do *not* need to rediscover the window upon auth connectivity interruption.
+```
+
+Both the systemd and k8s variants of the upgrade model need to assume that frequent agent crashes indicate an unhealthy
+state. Rather than adding additional health check mechanisms, agents that _know_ that they are unhealthy should perform
+a periodic exit to inform their updater. The primary metric of health will be auth server connectivity. If an agent
+cannot establish a healthy connection to any auth agent for greater than 5 minutes, it will exit.
```

Comment:
Being able to upgrade even if the agent is too outdated to talk to the control plane was an explicit design goal.  It's alluded to, but I'll try to update the intro to communicate that fact better.  One of the current blockers to cloud control plane upgrades is that some users ship devices/images with teleport installed that might not come online for a very long time, meaning that we want to be resilient to very old agents "appearing", after having been offline for long enough to potentially break compatibility in unexpected ways.

Furthermore, we consider permanent loss of access due to a bad update to be unacceptable, and therefor would prefer to attempt an update even when teleport is unhealthy rather than allowing a bugged update to permanently orphan a machine.

The decision to have an update be periodically attempted if teleport is unhealthy for a prolonged period of time emerges from the above concerns.

Reply:
Yeah that follows, thanks for the explanation